### ISSUE TITLE
HiOp test failure with MKL on OSX

### ISSUE THREAD

**Issue by ashermancinelli** [LLNL/hiop#381]
Building with MKL, GCC11 on osx gives

```console
. /opt/intel/oneapi/mkl/latest/env/vars.sh
CC=gcc-11 CXX=g++-11 cmake ..
make -j
ctest
...
95% tests passed, 1 tests failed out of 22

Total Test time (real) =  18.99 sec

The following tests FAILED:
	 16 - NlpDenseCons3_50K_mpi (Failed)
Errors while running CTest
```

Full log here: [nlpDenseCons_ex3.exe.log](https://github.com/LLNL/hiop/files/7764459/nlpDenseCons_ex3.exe.log)

**Comment by ashermancinelli** [comment]
CC @cnpetra, based on our convo offline.

**Comment by andersp** [comment]
I'm still having problems getting make test to pick up the mkl dylib on my mac laptop. I attach some log files.
[CMakeCache.txt](https://github.com/LLNL/hiop/files/7764852/CMakeCache.txt)
[LastTest.log](https://github.com/LLNL/hiop/files/7764862/LastTest.log)

**Comment by ashermancinelli** [comment]
Could you paste a full transcript of your shell session, along with anything in your `~/.bashrc` that might be getting in the way? For example, here is what works for me:
```console
mkdir build && cd build
cmake .. -DHIOP_USE_MPI=OFF
make -j 12
make test
```

And just for good measure:
```console
otool -L ./tests/testVector
./tests/testVector:
	/usr/local/opt/libomp/lib/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
	@rpath/libmkl_intel_lp64.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_sequential.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_core.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
```

**Comment by andersp** [comment]
(base) chebyshev:build petersson1$ otool -L ./tests/testVector
./tests/testVector:
	/usr/local/opt/libomp/lib/libomp.dylib (compatibility version 5.0.0, current version 5.0.0)
	@rpath/libmkl_intel_lp64.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_sequential.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	@rpath/libmkl_core.2.dylib (compatibility version 0.0.0, current version 0.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1311.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1200.3.0)

**Comment by andersp** [comment]
[bash_profile.gz](https://github.com/LLNL/hiop/files/7765011/bash_profile.gz)

**Comment by cnpetra** [comment]
anaconda tends to mess up user profile even after `conda deactivate` ... remove anything anaconda related from bash profile and try again...

**Comment by andersp** [comment]
Commenting out the conda stuff made no difference :-(

**Comment by andersp** [comment]
I noticed the all tests pass if I run 'ctest' instead of 'make test'.

**Comment by jwang125** [comment]
Hi, are you using the develop branch of hiop? This was an issue for me as well but it was fixed I think two weeks ago. https://github.com/LLNL/hiop/pull/365

**Comment by cnpetra** [comment]
> I noticed the all tests pass if I run 'ctest' instead of 'make test'.

Then I would say go ahead and use HiOp with your code. To confirm everything is fine, run any of the executables in `build/src/Drivers`

A potential fix is to not use intel mkl (also disable anaconda) and let cmake pick the blas and lapack that comes with laptop (at least my Mac comes with Accelerate versions of these). 

I would be happy to stop by your office and troubleshoot more. There is definitely something off on your system.

**Comment by andersp** [comment]
@jwang125 I tried switching to the develop branch but it has the same problem with 'make test'. The tests do pass with 'ctest', so I think my installation is ok.

**Comment by jwang125** [comment]
OK. Then I don't know what the issue is. It worked fine on Lassen for me. There is a different issue on summit. Could be related to the system you're using.

**Comment by andersp** [comment]
@jwang125 I'm on a mac :-)

**Comment by andersp** [comment]
I switched to building hiop with HIOP_USE_MPI=On. Now one to the tests (with ctest) fails:
The following tests FAILED:
	 16 - NlpDenseCons3_50K_mpi (Failed)
Errors while running CTest

It seems to be running on 2 mpi tasks and exits with this message:
Panic: minimum step size reached. The problem may be infeasible or the gradient inaccurate. Try to restore feasibility.
  16  2.0257870e+02 1.479e-08  7.537e-11  -9.00  1.000e+00  5.551e-17  54(?)
Couldn't solve the problem.
Linesearch returned unsuccessfully (small step). Probable cause: inaccurate gradients/Jacobians or infeasible problem.

**Comment by jwang125** [comment]
When MPI is on, the test is run with 2 ranks but the same problem. I just ran it on Lassen and it was working. No panic step. We can have a quick meeting if you want.

**Comment by andersp** [comment]
Sounds good. I will setup a webex meeting

**Comment by andersp** [comment]
@jwang125 what's you're email?

**Comment by jwang125** [comment]
wang125@llnl.gov

**Comment by nychiang** [comment]
@andersp @jwang125 please forward me the meeting invitation

**Comment by nychiang** [comment]
@ashermancinelli 
I am using macOS BigSur, gcc v11.2.0, openmpi and xcode default blas and lapack, and all the tests work fine for me.
However, I admit that I do observe this error couple of times, on mac/lassen/quartz/ascent. 
This problem is a hard one and it fails due to the different version of HiOp dependencies, e.g., MPI.
The rounding errors make the steps slightly different from the ones on the other platforms and finally activate this failure on some platform. The good news is that the solver can still converge, but only has this problem in the final iteration. 

Currently I don't have a good way to solve it, but just tune the hiop parameters to make the case converge on all the platform. Can you try to set tol=1e-6 and see if it can converge?

**Comment by jwang125** [comment]
Sorry we didn't see this message in time. We manged to compile but there are still some warnings from MacOS like version not compatiable. I will let Anders contact you if there are any new issues.

**Comment by andersp** [comment]
@nychiang sorry, I didn't see your message until now. Frank helped my sort out a couple of other issues with my code, so at least it builds now. I still need to figure out what flags to give the linker to pick up the MKL blas and lapack.

**Comment by nychiang** [comment]
@andersp probably -DLAPACK_LIBRARIES. I set -DLAPACK_LIBRARIES="-lessl -llapack" on lassen to use essl and lapack in the system search path.

**Comment by andersp** [comment]
The build process works perfectly if I remove the MKL directory from the DYLD_LIBRARY_PATH and instead use these flags:
cmake --install-prefix=/Users/petersson1 -DHIOP_USE_MPI=ON -DLAPACK_LIBRARIES="-framework Accelerate" ..
make
make test
make install

**Comment by andersp** [comment]
By the way, test #16 only failed once. Perhaps there is a memory bug, like an uninitialized variable lurking somewhere?

**Comment by nychiang** [comment]
@andersp Do you mean it works now, after you remove MKL? I guess it still fails when MKL is used. @ashermancinelli

**Comment by ashermancinelli** [comment]
quick note:
```
# this
cmake --install-prefix=/Users/petersson1
# should be this
cmake -DCMAKE_INSTALL_PREFIX=/Users/petersson1
```

Also, I'm not surprised to hear MKL seems to be the issue as [that's the original problem that motivated this issue.](https://github.com/LLNL/hiop/issues/381#issue-1087066814) If Apple's Accelerate library works and that suites you @andersp, I would recommend it. Most of our tests run with openblas, but we've tested with essl and cray's libsci as well. We will look into this @andersp, but hopefully this is enough to get you started. Cheers!

**Comment by cnpetra** [comment]
test 16 fails because of loss of accuracy in a linear solve. I've seen it happening with some lapack distributions. @andersp you can safely ignore it for now.

**Comment by cnpetra** [comment]
@nychiang, fyi, the loss of accuracy is caused by large dual multipliers. If you can test the elastic mode you're working on with nlp dense ex3, especially for 
```
./src/Drivers/nlpDenseCons_ex3.exe 50000
mpirun -n 2 ./src/Drivers/nlpDenseCons_ex3.exe 50000
```

**Comment by cnpetra** [comment]
closed by pr#408
